packages/core/src/lib/filter/v2/index.ts

```diff
@@ -0,0 +1,338 @@
+import { pipe } from "it-pipe";
+
+import { BaseProtocol } from "../../base_protocol.js";
+import { DefaultPubSubTopic } from "../../constants.js";
+
+import {
+  FilterPushRpc,
+  FilterSubscribeResponse,
+  FilterSubscribeRpc,
+} from "./filter_rpc.js";
+
+const log = debug("waku:filter:v2");
+
+const FilterV2Codecs = {
+  SUBSCRIBE: "/vac/waku/filter-subscribe/2.0.0-beta1",
+  PUSH: "/vac/waku/filter-push/2.0.0-beta1",
+};
+
+class Subscription {
+  private readonly peer: Peer;
+  private readonly pubSubTopic: string;
+
+  private newStream: (peer: Peer) => Promise<Stream>;
+
+  constructor(
+    pubSubTopic: string,
+    remotePeer: Peer,
+    newStream: (peer: Peer) => Promise<Stream>
+  ) {
+    this.peer = remotePeer;
+    this.pubSubTopic = pubSubTopic;
+    this.newStream = newStream;
+  }
+
+  async subscribe<T extends IDecodedMessage>(
+    decoders: IDecoder<T> | IDecoder<T>[],
+    callback: Callback<T>
+  ): Promise<void> {
+    const decodersArray = Array.isArray(decoders) ? decoders : [decoders];
+    const contentTopics = Array.from(groupByContentTopic(decodersArray).keys());
+
+    const stream = await this.newStream(this.peer);
+
+    const request = FilterSubscribeRpc.createSubscribeRequest(
+      this.pubSubTopic,
+      contentTopics
+    );
+
+    try {
+      const res = await pipe(
+        [request.encode()],
+        lp.encode,
+        stream,
+        lp.decode,
+        async (source) => await all(source)
+      );
+
+      const { statusCode, requestId, statusDesc } =
+        FilterSubscribeResponse.decode(res[0].slice());
+
+      if (statusCode < 200 || statusCode >= 300) {
+        throw new Error(
+          `Filter subscribe request ${requestId} failed with status code ${statusCode}: ${statusDesc}`
+        );
+      }
+
+      log(
+        "Subscribed to peer ",
+        this.peer.id.toString(),
+        "for content topics",
+        contentTopics
+      );
+    } catch (e) {
+      throw new Error(
+        "Error subscribing to peer: " +
+          this.peer.id.toString() +
+          " for content topics: " +
+          contentTopics +
+          ": " +
+          e
+      );
+    }
+
+    const subscription: PeerSubscription<T> = {
+      callback,
+      decoders: decodersArray,
+      pubSubTopic: this.pubSubTopic,
+    };
+    FilterV2.subscriptionsLog.set(
+      this.peer.id.toString(),
+      [
+        ...(FilterV2.subscriptionsLog.get(this.peer.id.toString()) ?? []),
+        subscription,
+      ] ?? [subscription]
+    );
+  }
+
+  async unsubscribe(contentTopics: ContentTopic[]): Promise<void> {
+    const stream = await this.newStream(this.peer);
+    const unsubscribeRequest = FilterSubscribeRpc.createUnsubscribeRequest(
+      this.pubSubTopic,
+      contentTopics
+    );
+
+    try {
+      await pipe([unsubscribeRequest.encode()], lp.encode, stream.sink);
+    } catch (error) {
+      throw new Error("Error subscribing: " + error);
+    }
+  }
+
+  async ping(): Promise<void> {
```

Comment:
The pattern here is `Promise<void>` and throwing if it fails. Not sure it's the best one, maybe returning a result would be better. Fine to keep it like this for now and look at usage.